src/osd/OSD.cc

```diff
@@ -3093,6 +3093,22 @@ will start to track new ops received afterwards.";
     pg_recovery_stats.reset();
   }
 
```

Comment:
I don't understand how this command is supposed to work.  Add brief, precise comments to the header declarations of convert_malformed_key and convert_malformed in the format:

```
/**
 * method_name
 *
 * Explanation of what method_name *does*, not how it works.  A reader should be able
 * to use this method correctly by reading this comment without needing to read the
 * implementation.
 */
```

If you need to elaborate on something in the implementation, do it in the definition body.

From what I can tell from the implementation, it looks like the goal here is to remove the malformed key and write back (for an ec pool) all 128 possible legacy keys.  What's the plan for ultimately removing the extras?  You could avoid this by passing in a set of all spg_t and pg_pool_t (for pg_num so you can run stable_mod) on the OSD.  Virtually every time, a particular PG will have only one shard present on the OSD, so you can take the hobject_t, map it to a pg, find the matching spg_t, and use that to determine the shard.  You could simply skip any objects in PGs that happen to have multiple shards present.

Reply:
Also, let's name it "fix_malformed_snapmapper_keys".  The description should reference the bug id and be more useful: "Scans the SnapMapper for malformed keys created by bug <bugid> during upgrade from <release> to <other release>."

Reply:
I'm worried about holding the osd lock during the scan as well.  While you have bounded the number of keys to remove, you're going to keep re-reading the SnapMapper up to that point.  Is it the case that the malformed keys always sort at the start of the region?  If so, that's an invariant you should explain in the implementation definition (not the header).

Reply:
The commit message needs to reference the bug and give a very brief description of the history and how this command is supposed to remedy it.

Reply:
You also need to check for and avoid overwriting the fixed key if it already exists.  During the original upgrade that caused the problem and now, it's possible that some of the objects had their SnapMapper keys rewritten and thereby fixed.  Overwriting them with the old value would be incorrect.